service/service.go

```diff
@@ -19,22 +19,106 @@ package service
 import (
 	"context"
 	"errors"
+	"fmt"
+	"time"
 
+	"github.com/golang/glog"
+	"github.com/golang/protobuf/proto"
 	pb "github.com/kubeflow/metadata/api"
 	mlmd "github.com/kubeflow/metadata/ml_metadata"
+	"github.com/kubeflow/metadata/schemaparser"
+)
+
+const (
+	wholeMetaPropertyName string = "__ALL_META__"
+	categoryPropertyName  string = "category"
+	artifactCategory      string = "artifact"
+	defaultTimeout               = 5 * time.Second
 )
 
 // Service implements the gRPC service MetadataService defined in the metadata
 // API spec.
 type Service struct {
-	mlmdClient mlmd.MetadataStoreServiceClient
+	mlmdClient       mlmd.MetadataStoreServiceClient
+	typeNameToMLMDID map[string]int64
+
+	// TODO: Once MLMD type system supports schema annotations, the following two fields can be obtained by quering MLMD types.
+	typeNameToID map[string]string
+	schemaset    *schemaparser.SchemaSet
 }
 
 // NewService returns a new MetadataService.
-func NewService(mlmdClient mlmd.MetadataStoreServiceClient) *Service {
+func NewService(mlmdClient mlmd.MetadataStoreServiceClient, schemaRootDir string) (*Service, error) {
```

Comment:
Service should not need to know anything about schema root dir. It should be a standalone service, whose only concern is to export endpoints for serving Artifacts, Executions and Events. We could just call a separate function to do pre-registration.

Also, pre-registration should happen the same way users would call the endpoint. A standalone function takes your JSON schema, converts it to ArtifactType, and just calls the endpoint to register the latter instead. That way, Service is not aware of JSON schema or its parser.

Reply:
I agree it is better to handle predefined & customized type the same way by calling an endpoint.

But now the problem is that we expose a different [ArtifactType](https://github.com/kubeflow/metadata/blob/master/api/api.proto#L43) from what current MLMD offers. So there needs another conversion between these two if we register types via endpoints. I'm trying to limit the work of this PR by handling predefined types first. It is an implementation details and we can move endpoint approach once we support customized types.

What do you think?

Reply:
I'm ok with registering types first before the other endpoint is ready. But I'm still not sure about this approach. I still see `category` and `artifactCategory`, which doesn't fit with the intended `ArtifactType` data model. Will you be removing all this later?

Reply:
The `category` is to indicate the predefined type is `artifactType` or `executionType`. It just makes the JSON schema self-explanatory. They won't be the properties. I added line 101.